tail/tail.go

```diff
@@ -118,7 +180,28 @@ func Tail(ctx context.Context, logger log.Logger, dir string, promMon *prometheu
 		logger:  logger,
 		monitor: promMon,
 	}
-	cpdir, k, err := wal.LastCheckpoint(dir)
+
+	var cpdir string
+	var k int
+	var err error
+	startTime := time.Now()
+	for {
+		cpdir, k, err = lastCheckpoint(dir)
+		if time.Since(startTime) > config.DefaultCheckpointInProgressPeriod {
+			return nil, ErrCheckpointInProgress
+		}
+		if strings.HasSuffix(cpdir, ".tmp") {
+			doevery.TimePeriod(config.DefaultNoisyLogPeriod, func() {
+				level.Warn(t.logger).Log(
+					"msg", "checkpoint in progress, waiting",
+					"checkpoint", cpdir,
+				)
+			})
+			time.Sleep(100 * time.Millisecond)
```

Comment:
I suggest making the sleep 1 second. (If the Go `time.NewTimer()` API were not so terrible, I'd wish to replace the Sleep with a single-use timer, but its not so easy so no worries, this means you might wait 1s after SIGTERM to really exit, no big deal.)

Reply:
Updated